src/VirtoCommerce.CatalogPublishingModule.Web/Controllers/Api/CompletenessController.cs

```diff
@@ -0,0 +1,283 @@
+        public async Task<ActionResult<CompletenessChannelSearchResult>> SearchChannel([FromBody]CompletenessChannelSearchCriteria criteria)
+        {
+            if (criteria == null)
+            {
+                criteria = new CompletenessChannelSearchCriteria();
+            }
+
+            var result = await _completenessService.SearchChannelsAsync(criteria);
+
+            return Ok(result);
+        }
+
+        /// <summary>
+        /// Create new channel
+        /// </summary>
+        /// <param name="channel">Channel</param>
+        [HttpPost]
+        [Route("channels")]
+        [ProducesResponseType(typeof(CompletenessChannel), StatusCodes.Status200OK)]
+        [Authorize(ModuleConstants.Security.Permissions.Create)]
+        public async Task<ActionResult<CompletenessChannel>> CreateChannel([FromBody]CompletenessChannel channel)
+        {
+            await _completenessService.SaveChannelsAsync(new[] { channel });
+
+            return Ok(channel);
+        }
+
+        /// <summary>
+        /// Update channel
+        /// </summary>
+        /// <param name="channel">Channel</param>
+        [HttpPut]
+        [Route("channels")]
+        [ProducesResponseType(typeof(void), StatusCodes.Status204NoContent)]
+        [Authorize(ModuleConstants.Security.Permissions.Update)]
+        public async Task<ActionResult> UpdateChannel([FromBody]CompletenessChannel channel)
+        {
+            await _completenessService.SaveChannelsAsync(new[] { channel });
+
+            return NoContent();
+        }
+
+        /// <summary>
+        /// Delete channels  
+        /// </summary>
+        /// <remarks>Delete channels by given array of pricelist ids.</remarks>
+        /// <param name="ids">An array of channels ids</param>
+        [HttpDelete]
+        [Route("channels")]
+        [ProducesResponseType(typeof(void), StatusCodes.Status204NoContent)]
+        [Authorize(ModuleConstants.Security.Permissions.Delete)]
+        public async Task<ActionResult> DeleteChannels([FromRoute] string[] ids)
+        {
+            await _completenessService.DeleteChannelsAsync(ids);
+
+            return NoContent();
+        }
+
+        private async Task<ActionResult<EvaluateCompletenessNotification>> EvaluateCompletenessAsync(string notifyType, string notificationDescription, Action<EvaluateCompletenessNotification> job)
+        {
+            var notification = new EvaluateCompletenessNotification(_userNameResolver.GetCurrentUserName(), notifyType)
+            {
+                Title = notificationDescription,
+                Description = "Starting evaluation..."
+            };
+            await _pushNotifier.SendAsync(notification);
+
+            job(notification);
+
+            return Ok(notification);
+        }
+
+        [ApiExplorerSettings(IgnoreApi = true)]
+        public async Task EvaluateCompletenessJob(string channelId, EvaluateCompletenessNotification notification)
```

Comment:
missed binding attributes for parameters may lead to an unexpected behavior such as null reference exception

Reply:
This method is non API (only for calling in background task), so binding attributes not needed.